src/librustc/traits/project.rs

```diff
@@ -1104,27 +1104,41 @@ fn assemble_candidates_from_impls<'cx, 'gcx, 'tcx>(
                     // an error when we confirm the candidate
                     // (which will ultimately lead to `normalize_to_error`
                     // being invoked).
-                    node_item.item.defaultness.has_value()
+                    false
```

Comment:
This looks very wrong to me, since it sounds like it'd allow assuming defaults inside the trait. For some reason that doesn't happen (the tests pass).

Reply:
@jonas-schievink 

Inside the trait, you won't be having a `VtableImpl` (which indicates a specific impl was selected), but a `VtableParam` (because of an "implicit where-clause").

Reply:
However, this does not correctly handle the case where an associated type default is overridden in a derived impl:
```rust
#![feature(associated_type_defaults, specialization)]

trait Foo {
    type Xyz = u32;
    fn foobar() -> Self::Xyz;
}

impl<T> Foo for T {
    fn foobar() -> Self::Xyz { 0u32 }
}

impl Foo for u32 {
    type Xyz = &'static u32;
}

fn main() {
    u32::foobar();
}
```

Reply:
However, according to the spec, the problem is somewhere else, because of
> To permit this is not a problem because `Foo for Vec<T>` is not further specializable since `Bar` in the implementation has not been marked as default.

Reply:
> However, according to the spec, the problem is somewhere else, because of

So the RFC says:

> If an implementation does make the associated type available for further specialization, then other definitions in the implementation may not assume the given underlying specified type of the associated type and may only assume that it is `Self::TheAssociatedType`.

which the snippet above clearly violates... but the more interesting question is "how do we ban it?"...

...For the sake of separate compilation (i.e. imagine `impl Foo for u32` is in a different crate), I think we need to say that `impl<T> Foo for T {` is legitimate but `impl Foo for u32 {` is *not*. Why? Because it overrides `type Xyz` which should not be up for specialization since it was not marked as `default` in `impl<T>` (if it did then `fn foobar` could not assume `type Xyz`'s definition). Moreover `fn foobar` here assumes `type Xyz`'s definition.

Reply:
Sure. If we go by the RFC, the default associated type `type Xyz = u32;` should only be specializable by the first impl that matches, not by any specializing impls (this can be checked at `2)` time because of the orphan rules), so `2)` should give an E0520 (`Xyz` specializes an item from a parent `impl`, but that item is not marked `default`).

I'll note that this contradicts the way associated fns work:
```Rust
#![feature(specialization)]

trait Foo {
    fn foobar()  { unreachable!() }
}

impl<T> Foo for T {
}

impl Foo for u32 {
    fn foobar()  { panic!() }
}

fn main() {
    u32::foobar();
}
```

Reply:
> also, if we do that, we have to decide whether associated consts should work like associated types or like associated fns.

The RFC applies to associated fns as well so how they work will need to be changed as well but not necessarily in this PR. The RFC says:
> This applies generally to any item inside a trait. [...]

which includes `fn`s and `const`s also.

Reply:
This is an open question in the design of specialization (#31844):
> Should default trait items be considered `default` (i.e. specializable)?

Specialization is unstable, so I don't think there's a real reason not to forbid respecialization of associated types and associated fns together. However, I *might* want to do this in an *earlier* PR, and assess crater impact.

Reply:
If we continue on this question: can you have `default` provided trait items that can be re-overriden?

```rust
#![feature(specialization)]

trait Foo {
    default fn foobar() -> bool { false }
}

impl<T> Foo for T {
}

impl Foo for u32 {
    fn foobar() -> bool { true } // OK
}

fn main() {
    println!("{:?}", u32::foobar()); // prints `true`
}
```

I think either option would be worth writing clearly in a "changes to specialization" section in RFC 2532, along with examples.

Reply:
> This is an open question in the design of specialization (#31844):
> 
> > Should default trait items be considered `default` (i.e. specializable)?

This is a bit of an orthogonal question; the main question is instead whether re-overriding is allowed as you point out later. I think it flows from the reference of the RFC and the spirit of it that re-overriding is banned. Instead, the idea is that `1)` in both of your snippets "lock in" what they do not explicitly allow for further specialization (with `default $item` in impls). Whether `trait` items must be prefixed with `default` to be specializable is not a particularly interesting question from this perspective and especially not from the POV of specialization groups as per in the "future possibilities" section of the RFC.

> I think either option would be worth writing clearly in a "changes to specialization" section in RFC 2532, along with examples.

It's a good idea to write down interesting points and decisions somewhere be it in the RFC or in the tracking issue. I think we can link to this discussion in a bullet point in the tracking issue description and then consider amending the RFC with new information at a later point. These notes will be useful when I (or someone else, but most likely not...) writes the stabilization report.